In this notebook, we are going to collect our data from Twitter. For this, you need to have fristly created a twitter developer account and after that, made an API which gives you keys for connecting. 
If you don't know how to do so, please follow this link.
When you create the API for your software, in your developer portal at Twitter website, you'll be able to reach the data details of your API including: API Key, API Key Secret, and Bearer Token
Next, you need to have some packages installed that you can gather data and from Twitter and do analysis on this data. In this notebook, we would describe and install all of these packages (if required), so, you won't need to install them for the next steps of the project.

The list of these packages are as follow:
* tweepy
* nlkt
* textblob
* collections
* matplotlib
* json
* pandas
* numpy
* pickle
* datatime
* pprint
* WordCloud

As mentioned-above, all of these packages are required for this project. However, for this file which is the first step, we would need only sum of them. So, we need to install them and use anything related which is required.
For the very first step, we need to install the packages for collecting data from Twitter. This package is tweepy.

Tweepy (https://www.tweepy.org/), is an easy-to-use Python library for assessing the Twitter API. To install we have:

In [22]:
!pip install tweepy

The next step is to gather data using this package and the API, you've made. So, we have:

In [31]:
import tweepy

Pay attention that for collecting data, we need a class made based on the tweepy.
This class has a name TwitterCollector. You would need to have the file of related to code of this class, in the place that you have this notebook. Then, you would be able to run the code. As soon as you upload this file, you need the following code:

In [34]:
from TwitterCollector import TwitterCollector

Moreover, as you are going to run functions that will collect data for a period of specific time, you'll ask them, you need another python package related to time and data. Most probably, you've installed it formerly. So, you should go for:

In [37]:
from datetime import datetime

Now, it is time to collect your data. You need your bearer_token and also an initialization for twitter_collector. 
Pay attention, 'bearer_token' should be the thing you've received from your tweeter developer account

In [42]:
bearer_token= r"AAAAAAAAAAAAAAAAAAAAAN7xhgEAAAAA5mDxdSxnPsiFrmzQ4R2Fvet71Qs%3DeAlmpOrJCjfAhrqhwnnbuZytWstuT1qdXZFLwPE6OZTo6ka2A6"
#initialize it:
tc= TwitterCollector(bearer_token=bearer_token)

Now, you are prepared to collect data. Just you need to tell about the queries you need. In this project, we are going to go for the word "Mahsa Amini", which is one of the greatest trends in Twitter till the end of September 2022.
The thing is that, we don't want retweets to be counted again. And we are just considering English Tweets not, Persian. Although the majority of tweets are in English.


In [43]:
query1= 'Mahsa Amini -is:retweet lang:en'

For collecting tweets, we have a function fetch_recent_tweets that the inputs include query, the number of tweets we want, and saving the results or not. So, we have:

In [45]:
recent_tweets_ma= tc.fetch_recent_tweets(query=query1, tweets_cnt=10000, save_result=True)

When you run this code, a json file will be created in the folder that you are running the code. You need this file for the next steps to do text mining.

Before going for text mining, we can have a better look on the data we collected in recent_tweets_ma:

In [47]:
type(recent_tweets_ma)

dict

In [48]:
recent_tweets_ma.keys()

dict_keys(['collection_type', 'collection_timestamp', 'query', 'tweet_cnt', 'tweets'])

We see that we have a dictionary with 5 different keys. The last key is 'tweets' which includes more data about a tweet. Let's have a look on it:

In [49]:
type(recent_tweets_ma['tweets'])

list

In [53]:
type(recent_tweets_ma['tweets'][0])

dict

So, we can see that 'tweet' is a list of dictionaries. Actually it includes all of our 10000 tweets and each member of this list is a dictionary which includes detail of each tweet.

In [54]:
recent_tweets_ma['tweets'][0].keys()

dict_keys(['entities', 'public_metrics', 'in_reply_to_user_id', 'possibly_sensitive', 'source', 'text', 'id', 'context_annotations', 'lang', 'author_id', 'referenced_tweets', 'created_at', 'edit_history_tweet_ids'])

The information about each tweet are the keys printed from the last line code. Again, each of these keys maybe another dictionary.
Indeed, the json file we get from this twitter api is a dictionary of dictionary of ... of dictionaries.

We can go further and do more with data to be familiar with. For example, we can get the unique author IDs:

In [113]:
uniq_author_id=[]
for tweet in recent_tweets_ma['tweets']:
    if tweet['author_id'] not in uniq_author_id:
        uniq_author_id.append(tweet['author_id'])

In [114]:
print(type(uniq_author_id))
print(len(uniq_author_id))

<class 'list'>
3306


As we see, we have 3306 unique author id. It means that the 10000 tweets that we've collected are made by 3306 different authors.

We may need to collect more data about this authors using their ID